File: c#/OttoCrypt.Demo/Controllers/OttoDemoController.cs
using System;
using System.IO;
using System.Text;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using IvanSostarko.OttoCrypt;

namespace OttoCrypt.Demo.Controllers
{
    public class OttoDemoController : Controller
    {
        [HttpGet]
        public IActionResult Index()
        {
            return View();
        }

        [HttpPost]
        [ValidateAntiForgeryToken]
        public IActionResult EncryptText(string plaintext, string mode, string? password, string? recipient_public, string? raw_key)
        {
            try
            {
                var opts = ParseOptions(mode, "encrypt", password, null, recipient_public, raw_key);
                var otto = new OttoCrypt();
                var (cipher, header) = otto.EncryptString(Encoding.UTF8.GetBytes(plaintext ?? string.Empty), opts);
                ViewBag.TextCipherB64 = Convert.ToBase64String(cipher);
                ViewBag.TextHeaderB64 = Convert.ToBase64String(header);
                ViewBag.ActiveTab = "text";
                return View("Index");
            }
            catch (Exception ex)
            {
                ModelState.AddModelError("text_error", "Encrypt error: " + ex.Message);
                return View("Index");
            }
        }

        [HttpPost]
        [ValidateAntiForgeryToken]
        public IActionResult DecryptText(string cipher_b64, string header_b64, string mode, string? password, string? sender_secret, string? raw_key)
        {
            try
            {
                var opts = ParseOptions(mode, "decrypt", password, sender_secret, null, raw_key);
                var otto = new OttoCrypt();
                var cipher = Convert.FromBase64String(cipher_b64 ?? "");
                var header = Convert.FromBase64String(header_b64 ?? "");
                var plain = otto.DecryptString(cipher, header, opts);
                ViewBag.TextPlainDec = Encoding.UTF8.GetString(plain);
                ViewBag.ActiveTab = "text";
                return View("Index");
            }
            catch (Exception ex)
            {
                ModelState.AddModelError("text_error", "Decrypt error: " + ex.Message);
                return View("Index");
            }
        }

        [HttpPost]
        [ValidateAntiForgeryToken]
        public async Task<IActionResult> EncryptFile(IFormFile file, string mode, string? password, string? recipient_public, string? raw_key)
        {
            if (file == null || file.Length == 0)
            {
                ModelState.AddModelError("file_error", "No file selected.");
                return View("Index");
            }

            var tempDir = Path.Combine(Path.GetTempPath(), "otto-demo");
            Directory.CreateDirectory(tempDir);
            var inPath = Path.Combine(tempDir, Guid.NewGuid().ToString() + "_" + file.FileName);
            var outName = file.FileName + ".otto";
            var outPath = Path.Combine(tempDir, Guid.NewGuid().ToString() + "_" + outName);

            try
            {
                await using (var fs = System.IO.File.Create(inPath))
                {
                    await file.CopyToAsync(fs);
                }
                var opts = ParseOptions(mode, "encrypt", password, null, recipient_public, raw_key);
                var otto = new OttoCrypt();
                otto.EncryptFile(inPath, outPath, opts);

                var bytes = await System.IO.File.ReadAllBytesAsync(outPath);
                System.IO.File.Delete(inPath);
                System.IO.File.Delete(outPath);
                return File(bytes, "application/octet-stream", outName);
            }
            catch (Exception ex)
            {
                ModelState.AddModelError("file_error", "Encrypt error: " + ex.Message);
                return View("Index");
            }
        }

        [HttpPost]
        [ValidateAntiForgeryToken]
        public async Task<IActionResult> DecryptFile(IFormFile encfile, string mode, string? password, string? sender_secret, string? raw_key)
        {
            if (encfile == null || encfile.Length == 0)
            {
                ModelState.AddModelError("file_error", "No file selected.");
                return View("Index");
            }

            var tempDir = Path.Combine(Path.GetTempPath(), "otto-demo");
            Directory.CreateDirectory(tempDir);
            var inPath = Path.Combine(tempDir, Guid.NewGuid().ToString() + "_" + encfile.FileName);
            var baseName = encfile.FileName.EndsWith(".otto", StringComparison.OrdinalIgnoreCase)
                ? encfile.FileName.Substring(0, encfile.FileName.Length - 5)
                : encfile.FileName;
            var outName = baseName + ".dec";
            var outPath = Path.Combine(tempDir, Guid.NewGuid().ToString() + "_" + outName);

            try
            {
                await using (var fs = System.IO.File.Create(inPath))
                {
                    await encfile.CopyToAsync(fs);
                }
                var opts = ParseOptions(mode, "decrypt", password, sender_secret, null, raw_key);
                var otto = new OttoCrypt();
                otto.DecryptFile(inPath, outPath, opts);

                var bytes = await System.IO.File.ReadAllBytesAsync(outPath);
                System.IO.File.Delete(inPath);
                System.IO.File.Delete(outPath);
                return File(bytes, "application/octet-stream", outName);
            }
            catch (Exception ex)
            {
                ModelState.AddModelError("file_error", "Decrypt error: " + ex.Message);
                return View("Index");
            }
        }

        [HttpGet]
        public IActionResult Keys()
        {
            try
            {
                var (secret, pub) = KeyExchange.GenerateKeypair();
                return Json(new {
                    secret_base64 = Convert.ToBase64String(secret),
                    public_base64 = Convert.ToBase64String(pub),
                    secret_hex = Convert.ToHexString(secret).ToLowerInvariant(),
                    public_hex = Convert.ToHexString(pub).ToLowerInvariant()
                });
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { error = ex.Message });
            }
        }

        private static Options ParseOptions(string mode, string op, string? password, string? sender_secret, string? recipient_public, string? raw_key)
        {
            mode = mode ?? "password";
            if (mode == "password")
            {
                if (string.IsNullOrEmpty(password)) throw new ArgumentException("Password required");
                return new Options { Password = password };
            }
            if (mode == "x25519")
            {
                if (op == "encrypt")
                {
                    if (string.IsNullOrEmpty(recipient_public)) throw new ArgumentException("Recipient public key required");
                    return new Options { RecipientPublic = recipient_public };
                }
                else
                {
                    if (string.IsNullOrEmpty(sender_secret)) throw new ArgumentException("Sender secret key required");
                    return new Options { SenderSecret = sender_secret };
                }
            }
            if (mode == "raw")
            {
                if (string.IsNullOrEmpty(raw_key)) throw new ArgumentException("Raw key required");
                return new Options { RawKey = raw_key };
            }
            throw new ArgumentException("Unknown mode");
        }
    }
}

File: laravel-test/app/Http/Controllers/OttoDemoController.php
<?php
namespace App\Http\Controllers;

use Illuminate\Http\Request;
use Illuminate\Support\Facades\Storage;
use Illuminate\Support\Str;
use Symfony\Component\HttpFoundation\StreamedResponse;
use IvanSostarko\OttoCrypt\Facades\OttoCrypt as Otto;
use IvanSostarko\OttoCrypt\KeyExchange;

class OttoDemoController extends Controller
{
    public function index()
    {
        // Render the demo blade
        return view('otto-demo');
    }

    public function encryptText(Request $request)
    {
        $request->validate([
            'plaintext' => 'required|string',
        ]);

        $options = $this->extractOptions($request, mode: 'encrypt');
        try {
            [$cipherAndTag, $header] = Otto::encryptString($request->input('plaintext'), $options);
            return back()->with([
                'text_cipher_b64' => base64_encode($cipherAndTag),
                'text_header_b64' => base64_encode($header),
                'active_tab' => 'text',
                'options_mode' => $options['__mode'] ?? 'password',
            ]);
        } catch (\Throwable $e) {
            return back()->withErrors(['text_error' => 'Encrypt error: ' . $e->getMessage()])->withInput();
        }
    }

    public function decryptText(Request $request)
    {
        $request->validate([
            'cipher_b64' => 'required|string',
            'header_b64' => 'required|string',
        ]);

        $options = $this->extractOptions($request, mode: 'decrypt');
        try {
            $cipher = base64_decode($request->input('cipher_b64'), true) ?: '';
            $header = base64_decode($request->input('header_b64'), true) ?: '';
            $plaintext = Otto::decryptString($cipher, $header, $options);
            return back()->with([
                'text_plain_dec' => $plaintext,
                'active_tab' => 'text',
                'options_mode' => $options['__mode'] ?? 'password',
            ]);
        } catch (\Throwable $e) {
            return back()->withErrors(['text_error' => 'Decrypt error: ' . $e->getMessage()])->withInput();
        }
    }

    public function encryptFile(Request $request)
    {
        $request->validate([
            'file' => 'required|file',
        ]);

        $file = $request->file('file');
        $origName = $file->getClientOriginalName();
        $tmpDir = 'otto-demo/' . Str::uuid()->toString();
        $inPath = $file->storeAs($tmpDir, $origName);
        $absIn = storage_path('app/' . $inPath);
        $outName = $origName . '.otto';
        $outPath = storage_path('app/' . $tmpDir . '/' . $outName);

        $options = $this->extractOptions($request, mode: 'encrypt');

        try {
            Otto::encryptFile($absIn, $outPath, $options);
        } catch (\Throwable $e) {
            return back()->withErrors(['file_error' => 'Encrypt error: ' . $e->getMessage()])->withInput();
        }

        return $this->downloadAndCleanup($outPath, $outName, [$absIn, $outPath]);
    }

    public function decryptFile(Request $request)
    {
        $request->validate([
            'encfile' => 'required|file',
        ]);

        $file = $request->file('encfile');
        $origName = $file->getClientOriginalName();
        $tmpDir = 'otto-demo/' . Str::uuid()->toString();
        $inPath = $file->storeAs($tmpDir, $origName);
        $absIn = storage_path('app/' . $inPath);
        $outName = preg_replace('/\.otto$/', '.dec', $origName) ?: ($origName . '.dec');
        $outPath = storage_path('app/' . $tmpDir . '/' . $outName);

        $options = $this->extractOptions($request, mode: 'decrypt');

        try {
            Otto::decryptFile($absIn, $outPath, $options);
        } catch (\Throwable $e) {
            return back()->withErrors(['file_error' => 'Decrypt error: ' . $e->getMessage()])->withInput();
        }

        return $this->downloadAndCleanup($outPath, $outName, [$absIn, $outPath]);
    }

    private function extractOptions(Request $request, string $mode): array
    {
        // Decide between password / X25519 / raw key modes from radio buttons
        $which = $request->input('mode', 'password');
        $opts = [];
        $opts['__mode'] = $which;

        if ($which === 'password') {
            $pw = (string)$request->input('password', '');
            if ($pw === '') throw new \InvalidArgumentException('Password required.');
            $opts['password'] = $pw;
        } elseif ($which === 'x25519') {
            if ($mode === 'encrypt') {
                $rcpt = (string)$request->input('recipient_public', '');
                if ($rcpt === '') throw new \InvalidArgumentException('Recipient public key required.');
                $opts['recipient_public'] = $rcpt;
            } else {
                $sk = (string)$request->input('sender_secret', '');
                if ($sk === '') throw new \InvalidArgumentException('Sender secret key required.');
                $opts['sender_secret'] = $sk;
            }
        } elseif ($which === 'raw') {
            $raw = (string)$request->input('raw_key', '');
            if ($raw === '') throw new \InvalidArgumentException('Raw key (32 bytes) required.');
            // Package will accept base64/hex/raw string
            $opts['raw_key'] = $raw;
        } else {
            throw new \InvalidArgumentException('Unknown mode.');
        }
        return $opts;
    }

    private function downloadAndCleanup(string $absPath, string $downloadName, array $cleanupPaths): StreamedResponse
    {
        return response()->streamDownload(function () use ($absPath, $cleanupPaths) {
            $fp = fopen($absPath, 'rb');
            while (!feof($fp)) {
                echo fread($fp, 8192);
                @ob_flush(); flush();
            }
            fclose($fp);
            // Cleanup temp files/dirs
            foreach ($cleanupPaths as $p) {
                @unlink($p);
            }
        }, $downloadName, [
            'Content-Type' => 'application/octet-stream',
        ]);
    }
}
